### DISCUSSION TITLE
Can company configured to fallback to other backends automatically when no matches found?

### DISCUSSION THREAD

**Discussion by adamcavendish** [company-mode/company-mode#1186] [Q&A] [answered]
Can I fallback to `dabbrev-code` for example, if company says that there's no match found?

**Comment by dgutov** [comment, marked as answer]
See #142.

The short answer is "no", but there's also a [third-party package](https://github.com/Wilfred/company-try-hard) (mentioned in the discussion linked above) which gives one alternative.

For manual invocation, you can also try the built-in command `M-x company-other-backend` and bind it to a key.

**Reply by yugaego** [reply]
I'd like to also *highlight* the following *indirect approaches* to this often asked (in all its variations) question:
- Fine-tuning grouped backends. [Start with `C-h v company-backends`.]
- Assigning particular [backend(s) to major-modes](https://github.com/company-mode/company-mode/wiki/Tips-&-tricks#configure-different-backends-for-different-modes).

**Reply by dalanicolai** [reply]
So can you suggest the best option please?

But let me first mention the following. I am using Spacemacs where, e.g. for `emacs-lisp-mode`, company-capf is configured as the first backend. Also, company-dabbrev(-code) is configured but `company-capf` `hijacks` most or all completions. So now the problem is that symbols that occur already within a buffer do not get suggested by autocompletion because company does not proceed to use the dabbrev backend. 

On the other hand, in `python-mode`, already occurring symbols do get suggested already by `company-capf` which uses `lsp-auto-completion-at-point`.

To make things work in `emacs-lisp-mode`, I can group `company-capf` with `company-dabbrev(-code)` to improve the situation, but I am not sure if that is the "correct/recommended" solution.

To make things more consistent (with `python-mode`), maybe the previous occurring words should also get suggested by `company-capf`/`elisp-completion-at-point` in `emacs-lisp-mode`? So then, should I report an issue for it at company-capf/emacs-bugs?

**Reply by yugaego** [reply]
> I can group company-capf with company-dabbrev(-code) to improve the situation, but I am not sure if that is the "correct/recommended" solution.

That's a fine solution (that can be tuned per exact setup/use case). It had been suggested ("recommended") a number of times, f.i. #484, #1041 (there're more but I didn't quick-find them, so you may look up for more examples).
Moreover, *rhetorically* speaking, if it works for you, doesn't it make it the "correct" solution (for you)?
 
I don't have sure/exact answers to the finalizing questions you listed, but they might be given later by (the maintainer) dgutov.

**Reply by dalanicolai** [reply]
Thanks for the quick answer. I was asking for it because I don't know how company's design is intended to work, and maybe you have some preferred design. But it seems that indeed it is recommended to group `company-dabbrev-code` with `company-capf` when the latter does not already provide the suggestions as provided by the former.

So I guess all is fine then, and the later questions have been answered implicitly.
Thanks again!

**Reply by dgutov** [reply]
> To make things more consistent (with python-mode), maybe the previous occurring words should also get suggested by company-capf/elisp-completion-at-point in emacs-lisp-mode? So then, should I report an issue for it at company-capf/emacs-bugs?

Maybe only in certain well-defined contexts? Otherwise the current implementation tries to make "smart" suggestions: offer functions where you need a function, offer variables where you need a variable, and so on.

**Reply by dgutov** [reply]
Simply mixing in all symbols from all buffers would mess with that goal.

**Comment by dgutov** [comment]
But the short answer is "no".
